src/download.ts

```diff
@@ -95,8 +112,10 @@ export class Download {
       await util.promisify(fs.unlink)(fileName).catch(e => {});
   }
 
-  _reportFinished(error?: string) {
+  async _reportFinished(error?: string) {
     this._failure = error || null;
+    for (const f of this._saveAsCallbacks)
```

Comment:
I don't see where this is assigned to.

Reply:
Oops. It was supposed to be this:

```
diff --git a/src/download.ts b/src/download.ts
index 1b3b67f0..8e6bf740 100644
--- a/src/download.ts
+++ b/src/download.ts
@@ -74,18 +74,20 @@ export class Download {
   }
 
   async saveAs(path: string) {
-    return new Promise(async (res, rej) => {
-      try {
-        const internalPath = await this.path();
-        if (internalPath === null)
-          throw new Error('Download not found on disk. Check download.failure() for details.');
-        if (this._deleted)
-          throw new Error('Download already deleted. Save before deleting.');
-        await util.promisify(fs.copyFile)(internalPath, path);
-        res();
-      } catch (error) {
-        rej(error);
-      }
+    return new Promise((res, rej) => {
+      this._saveAsCallbacks.push(async () => {
+        try {
+          const internalPath = await this.path();
+          if (internalPath === null)
+            throw new Error('Download not found on disk. Check download.failure() for details.');
+          if (this._deleted)
+            throw new Error('Download already deleted. Save before deleting.');
+          await util.promisify(fs.copyFile)(internalPath, path);
+          res();
+        } catch (error) {
+          rej(error);
+        }
+      })
     });
   }
```

I just tried the above patch and think it's now deadlocking. What did you have in mind to prevent the race?